DOC: Getting started guide for beginners (#460)
xorbitsai/inference: doc/source/getting_started/accessing_models.rst

.. _accessing_models:

================
Accessing Models
================

> .. note:: Suppose you have started the Xinference server endpoint at ``http://0.0.0.0:8001``. 

Reviewer: Would it be better to have xinference running at `http://127.0.0.1:9997` ?

Author: 127.0.0.1 is suitable for local deployment. 
What if I use "http://HOSTNAME:PORT" with placeholders instead?

Please refer to :ref:`Launching Models <launching_models>` guide for get your model running.

> Using Clients

Reviewer: ```suggestion
Using Xinference Client
```

=============

> LLM Model
> ---------

Reviewer: ```suggestion
Large Language Model
-----------------------
```

> When the Ability of the LLM model includes "chat", we can converse with it using the model's chat interface:

Reviewer: ```suggestion
When the abilities of the LLM include "chat," we can converse with it using the model's chat interface:
```

.. code-block:: python

   from xinference.client import RESTfulClient

>    client = RESTfulClient("http://0.0.0.0:8001")

Reviewer: ```suggestion
   client = RESTfulClient("http://127.0.0.1:8001")
```

>    model = client.get_model(model_uid)

Reviewer: Since the response below includes the model UID, I think it is better to get model with the model UID explicitly.

```suggestion
   model = client.get_model("f543d078-55f1-11ee-8fc0-0a28cc89f433")  # The model UID of the model you just launched.
```

   chat_history = []
   prompt = "What is the largest animal?"
   model.chat(
       prompt,
       chat_history,
       generate_config={"max_tokens": 1024}
   )

The response will look like:

.. code-block:: json

>    {

Reviewer: I generated another response that includes the usage.

```json
{
   "id":"chat3626defa-55f2-11ee-8f6c-0a28cc89f433",
   "object":"chat.completion",
   "created":1695020931,
   "model":"f543d078-55f1-11ee-8fc0-0a28cc89f433",
   "choices":[
      {
         "index":0,
         "message":{
            "role":"assistant",
            "content":"The largest animal in the world is the blue whale. It can grow up to 100 feet long and weigh up to 200 tons. It is the largest mammal on Earth and the largest living thing that ever existed. Blue whales are found in all of the world's oceans and are the only animals that can communicate with each other through a series of clicks and whistles. They are also the fastest animals in the world, able to swim at speeds of up to 35 miles per hour."
         },
         "finish_reason":"stop"
      }
   ],
   "usage":{
      "prompt_tokens":48,
      "completion_tokens":117,
      "total_tokens":165
   }
}
```

Author: Would you please make a change to this file so I can commit it on the webpage

     "id": "chatcmpl-8d76b65a-bad0-42ef-912d-4a0533d90d61",
     "model": "56f69622-1e73-11ee-a3bd-9af9f16816c6",
     "object": "chat.completion",
     "created": 1688919187,
     "choices": [
       {
         "index": 0,
         "message": {
           "role": "assistant",
           "content": "The largest animal that has been scientifically measured is the blue whale..."
         },
         "finish_reason": "None"
       }
     ],
     "usage": {
       "prompt_tokens": -1,
       "completion_tokens": -1,
       "total_tokens": -1
     }
   }

Embedding Model
---------------

To interact with Xinference's embedding model, i.e., inputting a text and getting an embedding from ``RESTfulClient``:

.. code-block:: python

   from xinference.client import RESTfulClient
   client = RESTfulClient("http://0.0.0.0:8001")

>    model = client.get_model(model_uid)

Reviewer: Better use `3ef99480-496f-11ee-9009-c2c8e4cad3f6` explicitly.

   model.create_embedding("write a poem.")

The response will be:

.. code-block:: json

  {
    "object": "list",
    "model": "3ef99480-496f-11ee-9009-c2c8e4cad3f6",
    "data": [
        {
          "index": 0,
          "object": "embedding",

>           "embedding": [-0.003699747147038579]

Reviewer: The value may not be correct:

```python
[-0.003699747147038579]
```

It should be a list of floats with a length equal to the dimension of the embedding model (e.g., 512).

Here's an example from OpenAI: https://platform.openai.com/docs/api-reference/embeddings/object

Author: I see, I forget to put ...

        }
    ],
    "usage": {
        "prompt_tokens": 37,
        "total_tokens": 37
    }
  }

> Using OpenAI Python SDK

Reviewer: Have you tested it?

Author: Copy from @aresnow1's PR

Author: Yep, this piece code has been tested locally

=======================

Xinference provides an OpenAI-compatible RESTful interface. Thus, you can also use the OpenAI Python SDK to
chat with the model via the service's endpoint:

.. code-block:: python

   import openai
   import sys

>    openai.api_base = "http://0.0.0.0:8001/v1"

Reviewer: ```suggestion
   openai.api_base = "http://127.0.0.1:8001/v1"
```
